### DISCUSSION TITLE
[OBSOLETED] OrcaSlicer printer profile discussion

### DISCUSSION THREAD

**Discussion by SoftFever** [OrcaSlicer/OrcaSlicer#1132] [General]
Continued printer profile discussion from #1100
To ensure we're all on the same page during the discussion, here are some terminologies used in OrcaSlicer profiles currently:

- `machine_model` refers to the specific machine model, such as Bambulab X1C or Voron 2.4. The corresponding JSON file follows this format: `Orca Arena X1 Carbon.json`.

- `machine` refers to the individual machine instance of a particular `machine_model`. It defines the nozzle size. The corresponding JSON file follows this format: `Orca Arena X1 Carbon 0.6 nozzle.json`.

- `filament` refers to the filament profile. The corresponding JSON file follows this format: `Arena ABS @Arena X1C.json`.

- `process` refers to the process or print settings. It is associated with a specific `machine`. The corresponding JSON file follows this format: `0.24mm Draft @Arena X1C.json`.

**Comment by SoftFever** [comment]
> I am sorry to be late to the party. This is something that I have wanted to bring up for a long time, but I was waiting to see what @SoftFever had for the goals of the project first.
> 
> **A general comment: we should not take this lightly as it breaks compatibility with Bambu Studio. I am fine with that. However, we should really take our time, discuss it at length, and try to get it right. It will be harder if it needs to be changed again. BBL printer users are going to struggle with it already, because they will not be able to go back and forth as easily.**
> 
> Instead of responding to each of the above comments, I am going to try and add my thoughts as one comment. Hopefully I can organize it to be understandable.
> 
> ### Profile Files
> Although the name of the file should be usable to identify the file, it should not be use by the slicer for anything else.
> 
> * all profile information should be stored in the file, including printer brand, model, firmware type (and supporting version?)
> * all information should be stored in a single file. BBL's two files per profile is not good because files can get out of sync or change independently
> * if files are going to have some sort of checksum to detect corruption, it should only issue a warning if it fails. It should be possible to hand edit profile files
> * dependencies should be specified using the identifying information in the parent file and not the name.
> 
> Because file names are not used to manage dependencies, the slicer will need to create its own index of existing profiles, which it already does I believe. The slicer should check the profile directory contents when started to update the index based on new or changed files.
> 
> The user should be notified when the slicer detects changed or new files.
> 
> Files should not be duplicated. I do not understand why BBL copies system profiles from the install directory into the appdata directory.
> 
> If the system files are going to be updated automatically, independent from updating the slicer, then
> 
> 1. the user must be notified when this happens
> 2. the previous system profiles should be moved to a backup/old directory in case the user has issues with the new versions
> 3. it would be nice if the user had the option of
>    4. seeing a list of files that changed
>    5. being able to see diffs of the changes
>    6. being able to see which of their custom profiles depend on the changed system profiles and how their custom profiles are affected.
> 
> ### Profile Directory
> BBL has an overly complicated directory structure, including using a printer directory as a sibling directory of the profile directory to hold information about AMS profiles.
> 
> I just realized why Bambu Studio is copying system profiles from the install directory to AppData. It only copies profiles for printers that you have selected during setup. I guess that makes sense.
> 
> I am looking at the appdata system directory for Orca and there is a BBL and a Custom. I do not think custom makes sense under system.
> 
> It does not make sense to me that OrcaSlicer/user has subdirectories, as if there are multiple users, when the AppData directory is for a specific user on Windows.
> 
> I would suggest that in AppData/Roaming/OrcaSlicer this directory structure is used.
> 
> * profiles
>   
>   * system
>     
>     * base  (this makes sense for filament and process, but maybe not machine)
>       
>       * machine
>       * filament
>       * process
>     * printer model
>       
>       * machine
>       * filament
>       * process
>       * assets
>     * printer model
>   * user
>     
>     * base
>     * printer model
>     * printer model
> 
> For the individual files each should have a name that includes brand, model, nozzle size. Each model should have a base profile that includes all information for that model of printer, and then derived profiles for each nozzle size. This allows a printer to be defined once, and different nozzles to be added with minimal extra information.
> 
> I would really like to see the printer model have a profile and nozzle profiles to be a separate type of profile, or at least a machine profile without any of the settings specific to the machine specified.
> 
> In addition, there can be multiple profiles for a single brand/model, for example someone might have two Prusa MK3S printers, but have klipper on one of them. The printer profile/definition specifies the extruder area, the bed size and shape, but then there are two derived profiles that specify the machine limits because running klipper can change those so they are different than when running prusa firmware.
> 
> ### UI
> I posted a long issue for Bambu Studio about scoping [bambulab#1838](https://github.com/bambulab/BambuStudio/issues/1838) and how Project should be included in the scope hierarchy along with all the plates and objects. The most important part of that was to allow plates to also have a printer, nozzle, filament, and process profile selected. This makes plates much more useful.
> 
> Given that Orca is supporting so many other printers, I think one of my suggestions is even more important. If a user has multiple printers and they are different brands and models, the current selection of "printer" is not manageable. The user does get to use the setup wizard and restrict the set of printers that they have so that only those profiles will be shown. That could still leave a list in the drop-down showing dozens of options. Just because profiles are stored with separate file for each printer nozzle combination, and have file names like that, does not mean that the user should have it presented that way.
> 
> A better user experience would be to select the printer model to be used for this project (and also for each plate if desired). Change from two drop-downs to three.
> 
> * Printer Model: after the user makes a selection the other two drop-downs have their lists constrained to compatible choices. Because the use will have chosen available printers during setup, this will be a short list, possibly only a single entry
> * Nozzle Profile: with only the nozzle profiles available for the selected printer model this should be a short list. Having already selected the printer model, the nozzle profile choices do not need to show brand and model for the printer again and will be shorter and easier to distinguish.
> * Plate Type: this is only the plates allowed for the printer/nozzle profile (Note: The label for the drop-down should be "Plate Type" and not "bed type". The bed is the heated part that is under the plate.
> 
> So an example for my currently printing job: Printer: BBL P1P Nozzle Profile: .4mm - fast start Plate Type: Textured PEI
> 
> Note: that adding nozzle into the nozzle profile name is not required, neither is adding plate into the plate type.
> 
> Showing how this fits into the browser:
> 
> ```
> Project  Printer: BBL P1P  Nozzle: .4 - fast start  Plate: Textured PEI  Filament: Hatchbox PLA
>   - Plate 1  Printer: BBL P1P  Nozzle: .4 - standard  Plate: Textured PEI
>     - Object 1
>   - Plate: Optional Parts  Printer: Prusa MK3S  Nozzle: .8 Plate: Smooth PEI  Filament: Generic CF
>     - hanger
>     - post
> ```
> 
> That should be enough to present the general idea and start a discussion. (We should have the UI discussion in a separate issue or discussion.)
> 
> ### Profile Management
> I think that profiles should be moved to a separate GitHub repo under the Orca project. I am going to create a separate issue for this and explain.
> 
> **I really think we should break this down into several focused discussions so that each part can be clarified before moving forward. Changing profiles is very significant and completely breaks compatibility with Bambu Studio.

@julie777 Move the discussion here

**Reply by SoftFever** [reply]
> Given that Orca is supporting so many other printers, I think one of my suggestions is even more important. If a user has multiple printers and they are different brands and models, the current selection of "printer" is not manageable. The user does get to use the setup wizard and restrict the set of printers that they have so that only those profiles will be shown. That could still leave a list in the drop-down showing dozens of options. Just because profiles are stored with separate file for each printer nozzle combination, and have file names like that, does not mean that the user should have it presented that way.
> 
> A better user experience would be to select the printer model to be used for this project (and also for each plate if desired). Change from two drop-downs to three.
> 
> * Printer Model: after the user makes a selection the other two drop-downs have their lists constrained to compatible choices. Because the use will have chosen available printers during setup, this will be a short list, possibly only a single entry
> * Nozzle Profile: with only the nozzle profiles available for the selected printer model this should be a short list. Having already selected the printer model, the nozzle profile choices do not need to show brand and model for the printer again and will be shorter and easier to distinguish.

This makes a lot of sense, and I completely agree.

**Reply by SoftFever** [reply]
> all information should be stored in a single file. BBL's two files per profile is not good because files can get out of sync or change independently

I haven't got time to go through your whole proposal, here is my initial thought regard this.
It will make the profile reusable very difficult if all are stored in one file.
For example, it's commonly practiced to create filament profiles shared by multiple machines.

**Reply by fr3ak2402** [reply]
### UI
As @SoftFever said, the thing with the selection of the printers in the setup window of the printers, I also think it makes sense. Since the list of printers is also getting longer. I will add more printers in the future.

------------------------------------------
### Profile Directory
Now a short explanation, how this works with the system files and custom files and which role the AppData folder has thereby.

As you have already noticed, the slicer only copies the profile files into the AppData if they have also been set up. Furthermore, this also happens when you change a process, filament or machine. A new file is written, which has a link to the original file. This link is written in the JSON parameter `inherits` in the file. All parameters that you have changed in the system file will only be added to the custom file. For example, if you select a custom process in the slicer's UI, it also loads the values from the system file, because they are linked to each other.

The files are copied from the ressources folder to the AppData folder when a printer is created, this also has the reason that when an update is performed on the slicer, the files in the ressources folder of the printer may have changed. Then it compares the files in the AppData folder with those from the resources and updates them automatically, so that the user always uses the current version.
The AppData folder is used for this purpose.

In a Windows operating system there are several AppData folders, one directly in the root directory of the OS and one in the user folder. The slicer always maintains these in the AppData/Roaming which is under the user folder.

The AppData folder is a folder where all programs should write their files. This folder is also hidden from the user and can be made visible via a setting. In addition, in this folder no management of files via the user, this should only happen via the program. 

The user folder in `C:\User\{username}\AppData\Roaming\OrcaSlicer\user` has its good and meaningful use. Moreover, this is needed.

By default, everyone has a default folder in this folder, in which the slicer copies all files of a printer into subfolders. 

There is also a case where an additional folder is created in the user folder. As you may have noticed, you can log in to Bambulab Studio as well as OrcaSlicer with your Bambulab account. 

Everyone who has an account with Bambulab has a userid in the form of a number. The account also has a cloud storage, in which all the profiles that you have set up are stored.

When you log in to slice with this account, it creates a folder with a Bambulab userid. In this folder copies everything you have set up and will set up again. After that it will be synchronized to the cloud. 
In addition, it is also possible to disable the cloud synchronization, if this happens, he writes again all files in the deault folder. When you turn the synchronization back on, all files are copied back to the userid folder by default and synchronized.

Therefore, the slice can distinguish before the file comes from the ressoucres folder or from the cloud.
The slicer thus manages the file in separate folders, this is also very useful for me.

Your restructuring of the Profile folder I did not quite agree, because this brings a very large amount of changes in the slicer itself, in addition then all files as the cloud and locally together in a folder are managed. Then the synchronization between client and cloud also no longer works properly, since all files are loaded into the cloud. The option that one has now, that not all files of the profiles are stored in the cloud would then be omitted.
Also, the AppData folder is not a management folder of files created and intended to be placed there by the user.

This becomes very cumbersome for mac users, where the data is written to the `library` folder. With each change, the admin password would then have to be entered, despite admin rights, as this is a security function of macOS.
In addition, this can also lead to problems when using the slicer further. So I advise against doing this.

But I think some things from your folder structure are very good. Creating an assets folder in the profiles in the resources folder is one thing I would do as well. This also makes the folder look tidier.

------------------------------------------
### Profile Files
Unfortunately, I have to agree with @SoftFever when you suggest managing everything in one file. This would not affect the reusability of any of the profiles.

I also know that PrusaSlicer maintains everything in a file. But I find the way the Bambulab slicer does that in separate file first clearer than the PrusaSlicer.
In addition, I have already installed over 40 printers in the OrcaSlicer by creating the profiles.

Since the PrusaSlicer, Bambulab Studio and OrcaSlicer are based on each other, this does not mean that all slicer profiles have to be compatible with each other. In addition, all three slicers are independent slicers with their own functionalities.

I also created many printer profiles for Cura and PrusaSlicer. Maintaining the profiles in the Bambu Studio/OrcaSlicer is much clearer than with the other two slicers mentioned.

I can also judge this because **I created most of the printer profiles for the OrcaSlicer**...

We have now adjusted some things so that all profiles follow the same name conversations (except for the BBL printers), so that the profiles are clearer in this respect and other people can create profiles more easily. As well as having a template that you can use as a guide.

In addition, BambuLab has also written a contribution in their wiki for creating (**Add 3rd Party Printer into Bambu Studio**). There you can also see how the structure of the profiles is and you can also download a complete example.

https://wiki.bambulab.com/en/software/bambu-studio/add-third-party-printer-in-studio

Only the whole file names had to be adapted to the specific name conversion. We can also create a contribution on this topic in our wiki on GitHub, which is also planned.

Why should the slicer notify me when something has changed? The slicer also ensures that all my profiles are up to date.

The user is not interested if something changes. Since all his profiles not affected by it, only the values ​​that he didn't adjust. For example, if a new function is introduced in a printer and there is a change in the G-code, then this should simply be changed.

In addition, all changes will be announced when the slicer is released. This could be edited a bit better to make it clearer. Or in the slcier a window appears in the form of a change log after the update.

Since only the changes are made if a slicer version is also used.

As I explained above, how the slicer compares and manages the files from the AppData with those from the resources folder.

**Comment by julie777** [comment]
> Continued printer profile discussion from #1100 To ensure we're all on the same page during the discussion, here are some terminologies used in OrcaSlicer profiles currently:
>     * `machine` refers to the individual machine instance of a particular `machine_model`. It defines the nozzle size. The corresponding JSON file follows this format: `Orca Arena X1 Carbon 0.6 nozzle.json`.
> 

I would like to propose a small change. You say that `machine` refers to an individual instance. I think that is reasonable. But I think an instance is one of the four P1Ps that a print farm has and is sending prints to.

I am in complete agreement that a given machine model has many settings that are based on the physics of the hardware and should only be specified once for that model. I also agree that in addition to the unchangeable part of the machine there is another set of configuration settings that apply to the machine_model based on how it is configured. These include nozzle size, build plate type, possibly fans installed (such as aux fan or chamber fan). I propose that this be called `machine_configuration` instead of `machine`. This is a much more descriptive name. `machine is so generic that its meaning is not obvious. Imagine looking at a directory named profiles and seeing
  - machine
  - machine_model
  - process
  - filament
What does machine mean? 

I have long wondered about the categorization of settings and which groups should have profiles and which profiles should have not only inheritance dependencies, but dependencies on another selection, such as printer model.

Prusa Slicer has printer, filament, and print profiles that can be saved.
BBL changed it to printer, filament and process.

I think there is at least one more group of settings and that is the settings that are applied when the print is sent. This group is currently very limited, but there could be more settings possible. 

### Dependencies

I am going to use dependencies to mean "this requires that" as opposed to using the inheritance of profile settings.

Given a very large number of profiles (group of related settings) the slicer needs to be able to help guide the user and reduce the complexity. For example, in the slicer with the current UI, if I have 3 different printer models and 3 nozzle sizes for each, and some variations, such as different machine start G code, the drop-down might be showing more than 30 choices. One of the ways that choices was reduced was by moving build plate type out of the printer settings.

Separating the printer model from the printer configuration (nozzle size, plate type, enclosure) makes a huge difference here. After selecting the printer model, the list only has the nozzle sizes for that model. The rest of this solution should be obvious.

Another part of the simplification by using the dependencies is filament selection. If a given printer does not have an enclosure and cannot print ABS, then that should remove ABS from the filament choices. (I just realized that enclosed is another printer_config setting.)

Given the current ecosystem, I suspect that most of the process profiles are tied to a printer model. This is another dependency that needs to be used by the slicer when the user has multiple printers of differing models.

#### A different perspective

I just want to throw the into the mix. If I have a print farm with multiple machine models, some enclosed, and some of the same models with different nozzle sizes or built plate types, how would I optimally slice and print models?

I need to slice and then print a large number of models. Of course the first step is to slice. But do I want to start slicing by selecting a printer model? I do not think so. I think what I want to do is specify the necessary configuration for the model being printed. I want to choose:
  - nozzle size
  - type of build plate, which I might leave blank because is does not matter
  - type of filament(s) and color(s)
  - etc.

Next I want to set the parameters that control how it will be sliced. (These are called the process, but they may be a mixture of slicer settings and print settings. (I can understand why Prusa has print settings and BBL changed to process settings. Most of the settings control the slicing process, but there is a small difference from the user perspective where some settings seem more like printer and others seem more like slicer.)

When slicing is complete, the next step is to add to the print queue. Given the original lack of selection of printer model, the print manager can choose a printer based on characteristics of the printer, and additional information specified when the job is queued, such as priority. So for example, a print specifying ABS can only be printed on a printer that has an enclosure. A print with a very high priority should be sent to the fastest printer available.

I do realize that this is more of imagining the future, but is does have some affect on the discussion about profiles, profile types, and such.

**Reply by fr3ak2402** [reply]
@julie777 
Because we are talking about a slicer that is intended for 3D printers nozzle, so what else should be meant by machine than the printer?

So all our third party profiles have no dependencies to other printers, all printers of one manufacturer are independent and have their own processes. All of a manufacturer's printers are managed in a printer family and then also managed in the manufacturer's JSON file. In addition, the printer manufacturer also has its own filaments, which are then assigned to all the printers of the manufacturer.

So people who run a printer farm will use the Repetier server because it supports functionality for sending a file, such as launching, etc., to multiple printers at the same time.

It is also possible to send a file from OrcaSlicer to a Repitier Server. Therefore, this can also be controlled directly via the Repitier Server.

--------------------------------------------
When we submit a printer with a nozzle size of 0.4, we are shown processes to that nozzle size. Also, if a user creates processes for a 0.6 nozzle based on the 0.4 nozzle, they will also be displayed in the 0.4 nozzle. As this is not intended, a profile for a 0.6 nozzle should be maintained in the profile if the printer has a 0.6 nozzle at all. If the printer does not support a 0.6 nozzle at all, it should not have any processes for it.

We have not maintained all printers with a 0.6 nozzle, as this was cumbersome before I adapted all profiles. But this will be done in time for all printers, so that they also have other nozzle diameters.

**Comment by julie777** [comment]
@fr3ak2402 
I have already commented in much detail in another discussion, in which I mentioned you, so I will keep it brief here.

I do not think that I conveyed my meaning well enough. Sorry.

A printer model needs a profile that defines all of the characteristics of that model, which includes possible acceleration, which firmware it uses, and the flavor, and and gcode customization. These settings are specific to the printer model.

In addition, a printer can be configured differently for different jobs. I might change nozzle size, or change the build plate, etc. As you mentioned, only profiles for a .6 nozzle size should be shown if that is all the user has. What if the user has three P1Ps and one of them has a .6 nozzle and two of them have .4 nozzles. 

In addition, there will probably be users like me who with only a P1P with the .4 nozzle still have several different profiles for the printer because that is the only way to have multiple types of start up G code for the machine currently. 

I feel like I am still not explaining well. Part of the problem is that there are so many settings and currently they live in only three types of profiles, but there are many dependencies among the various settings and profiles. I would really like to discuss that as a separate topic and see if there is a better way that what someone came up with more than 10 years ago.

One example of an improvement that I would like to consider, is making more choices based on compatibility. BBL started trying to do some of this with bar codes on filament and build plates. That is not always possible, but other things like having a list of printers in the application and setting the nozzle size in the app for that printer when it is changed would be useful. 

Please read my last comment in #1156 and let me know if it makes any sense. 

I was planning to start working on a profile management app, which is separate from the slicer, but could be invoked from the slicer. I have already started collecting profiles as part of creating a profile library. I would like to contribute, but I am not sure if my ideas are compatible with yours. It might just be poor communication on my part, or maybe I am looking further into the future. Regardless, I acknowledge your contribution and thank you for it!

**Reply by fr3ak2402** [reply]
Yes, I understand the direction they want to take the OrcaSlicer. I think in many places it's just a bit too far thought out or too fast, so it's a bit overwhelming.

In some places it was a bit incomprehensible for us, but that was because of the amount of information or changes they proposed.

I don't think it's a bad idea to group certain parts of the slicer into a kind of app.

Like the Printer Profile Configurator thing, for example....

--------------------------------------------------------
I would suggest that I first rebuild all the profiles so that they have a uniform system in terms of files, etc. In addition, all of them should have a buildplate model such as texture. And also more than just a 0.4 nozzle setting.

I am working on this at full speed, and this should also flow into version V 1.6.4.

Then I would start with a Printer Profile Configurator, which gives us the possibility to easily create a new profile for a printer based on a profile, which writes the appropriate files etc. This would first of all improve the creation and make it easier for the user.

Once we have that, then we can implement other things.
So what I'm saying is we should always change one thing as much as we can. Because then we can plan and implement the other change.

What we should avoid at all costs is to start making changes in all parts of the slicer, otherwise this will end in a disaster.
It will take us a little longer to get everything done, but there are not many of us really working on the slicer project. What we must not lose sight of. So we have to plan a little differently than a company with many employees.

I would also say that a configurator or manager or library would have to be an app that would be called from the OrcaSlicer.

We will definitely find a way to put our goals or ideas into an app.

I've already developed a lot of things, but we can still incorporate functions of your ideas.

------------------------------------------------------------------
Should we then be ready, then we can gladly incorporate the next topic.

**Reply by julie777** [reply]
@fr3ak2402 
Thanks for responding and helping to clarify things.

**Comment by macdylan** [comment]
Any updates? or we can continue to commit with the previous specification...

**Reply by SoftFever** [reply]
@macdylan 
Sorry for not being able to follow up this topic.
A little busy  :)

Yes, we can continue with previous specfication.
I will comment in PR directly.

**Comment by SoftFever** [comment]
> Any updates? or we can continue to commit with the previous specification...

**Comment by FFengIll** [comment]
Is it possible to export all things under a machine (current version only export selected machine/filament/process)
